src/utils.c

```diff
@@ -746,3 +746,181 @@ gboolean r_fakeroot_cleanup(const gchar *envpath, GError **error)
 
 	return rm_tree(tmpdir, error);
 }
+
+gboolean semver_parse(const gchar *version_string, guint64 version_core[3], gchar **pre_release, gchar **build, GError **error)
+{
+	g_autofree gchar *version_copy = g_strdup(version_string);
+	gchar *build_pos = NULL;
+	gchar *pre_release_pos = NULL;
+	g_auto(GStrv) version_parts = NULL;
+
+	if (pre_release != NULL)
+		*pre_release = NULL;
+
+	if (build != NULL)
+		*build = NULL;
```

Comment:
This will leak memory, if this already points to a string, as in the tests.

Take a look at https://github.com/rauc/rauc/blob/master/src/mount.c#L115, there we use `g_return_val_if_fail` to check that the output parameter is either `NULL` or points to `NULL`. Later on success, `g_steal_pointer` is used to write the result back if needed: https://github.com/rauc/rauc/blob/master/src/mount.c#L221. On error, the string is freed via `g_autofree`.

Reply:
so this way a pointer *has* to be passed in - or its an error
but here my intention is to give the user the option to pass in NULL as in "don't care about that returned value"

Reply:
Ah, right. The it should look like the `error` checks we have in many functions:
```g_return_val_if_fail(pre_release == NULL || *pre_release == NULL, FALSE);```

Reply:
v4 works like i had initially intended - and does not seem to leak... so ok? (-:

Reply:
I'd prefer to use a consistent pattern of checks in RAUC. :)

Reply:
Ok, changed it to the `g_return_val_if_fail` with v4.1